### PR TITLE
Change docs for feval

### LINKED ISSUE #4759: [docs] [python] custom metric function in `lgb.train()` interface should not refer to the the passed dataset a `train_data`
## Summary

This issue proposes changing the documentation for `feval` in `lgb.train()`, in the Python package, replacing `train_data` with `evaluation_dataset`.

## Motivation

`lgb.train()` allows providing custom evaluation metrics defined in Python functions. According to the docs for `lgb.train()` ([here](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train)), such functions should have the following signature.

> Should accept two parameters: preds, train_data, and return (grad, hess).
>
> **train_data: Dataset**
>    The training dataset.

I think it's confusing to refer to this as "The training dataset". In my opinion, that gives the impression that custom evaluation functions will have access to the data being used for training, which is not true. Instead, each custom evaluation function is provided *a Dataset that evaluation has been request for*, which might be either the training data or a separate validation dataset.

https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py#L296-L299

https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3240-L3255

https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3271-L3272

## Description

Specifically, I'm proposing the following changes:

* change the following to `evaluation_data`, not `valid_data` https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/basic.py#L3254-L3255
* change the following to `evaluation_data`, not `train_data` https://github.com/microsoft/LightGBM/blob/da98f24711a2faab17f94e5b2a636e6609c93fa6/python-package/lightgbm/engine.py#L88-L89

## References

Noticed this while working on https://github.com/microsoft/LightGBM/issues/4679#issuecomment-955872028.

### PR DESCRIPTION
**Pull request by akshitadixit**
Changed `train_data` to `eval_data` in engine.py as well as basic.py

closes #4759

### CODE REVIEW

**Review by shiyu1994 — APPROVED** [review]
Thanks for doing this. Approved according to previous discussions.
